### DISCUSSION TITLE
Support Google Tag Manager and GA4 Data Layers

### DISCUSSION THREAD

**Discussion by brobatr** [adobe/aem-core-wcm-components#2565] [Ideas]
hi everyone,  I've been working on a feature to support push to Google data layers.  This is advantageous because these are much more widely used than the ACDL, so making the work done on datalayers so far of use to many more customers.  I think this is an excellent low-hanging fruit to delivering more from CC.

The changes are quite minor.  It involves:

- allow configuration of datalayer name (GDL normally called dataLayer)
- allow conditional use of gtag() or push().  These result in different data formats in the DL (array vs object).  GA4 code will use gtag, GTM code will use push
- testing and documentation

Unlike Adobe solutions which are datalayer agnostic, Google solutions do require a GDL and they create one if it doesn't exist.  There is no need to include Google code to instantiate a datalayer.  Google uses a helper object to add EDDL functionality to an existing array and the helper processes historical pushes to that array, i.e. no timing issues.  There is no conflict between consumers.

I have almost completed implementation and tested on both GA4 and GTM.  Events and data are showing correctly.

There is quite a lot of confusion out there about whether or not our products need the ACDL and what the difference is to the GDL (no and not very much).  This feature will help clear that and reassure customers that the Core Components tracking will work in a Google ecosystem.

There is an additional advantage that pre-sales demonstrations of AEM with GA4/GTM integration become much simpler.

I would appreciate feedback as I have been asked for confirmation that the feature will actually be accepted to get the funding to complete it.

**Comment by vladbailescu** [comment]
Hey, Rob!

I see no reason for not accepting the feature, provided that:
- the feature is properly implemented and documented
- meets quality criteria
- can be easily turned on or off
- is backwards compatible and will not break existing websites or customisations

We look forward to a more technical/in-depth description of the solution (or a PR).

Thank you!

**Reply by brobatr** [reply]
thanks @vladbailescu !

**Comment by brobatr** [comment]
hi @vladbailescu, finally... https://github.com/adobe/aem-core-wcm-components/pull/2613 and particularly the readme https://github.com/adobe/aem-core-wcm-components/blob/34fd8fbbce9d1078ac69a752af862fb53a0b6dba/DATA_LAYER_INTEGRATION.md .  

Build is failing because of baseline but otherwise ready for an initial review and feedback.  Is this comment enough to request that?

**Comment by brobatr** [comment]
This feature is on hold pending merge of this PR https://github.com/adobe/aem-core-wcm-components/pull/2624 to fix this bug https://github.com/adobe/aem-core-wcm-components/issues/2620
